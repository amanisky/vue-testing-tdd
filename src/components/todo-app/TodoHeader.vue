<template>
  <header class="header">
    <h1>待办</h1>
    <input
      class="new-todo"
      data-testid="new-todo"
      autofocus
      placeholder="需要做什么？"
      @keyup.enter="onEnter"
    />
  </header>
</template>

<script>
export default {
  name: 'TodoHeader',

  methods: {
    onEnter (e) {
      const text = e.target.value.trim()
      if (!text.length) {
        return
      }
      this.$emit('new-todo', text)
      e.target.value = ''
    }
  }
}
</script>
