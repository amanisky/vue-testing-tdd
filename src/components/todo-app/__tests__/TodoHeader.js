import { shallowMount } from '@vue/test-utils'
import TodoHeader from '@/components/todo-app/TodoHeader.vue'

describe('TodoHeader', () => {
  let wrapper = null

  beforeEach(() => {
    wrapper = shallowMount(TodoHeader)
  })

  test('new todo', async () => {
    const input = wrapper.findComponent('input[data-testid="new-todo"]')
    const text = 'play'
    await input.setValue(text)
    await input.trigger('keyup.enter')
    // 期望发射 new-todo 事件
    expect(wrapper.emitted('new-todo')).toBeTruthy()
    // 期望发射 new-todo 事件，携带的数据是 play
    expect(wrapper.emitted('new-todo')[0][0]).toBe(text)
    // 期望 input 输入框被清空
    expect(input.element.value).toBe('')
  })

  test('new todo with empty text', async () => {
    const input = wrapper.findComponent('input[data-testid="new-todo"]')
    const text = ''
    await input.setValue(text)
    await input.trigger('keyup.enter')
    // 期望不发射 new-todo 事件
    expect(wrapper.emitted('new-todo')).toBeFalsy()
  })

  // 组件开发完成并且稳定（样式和结构不再需要调整），加上快照测试，避免无意间修改
  test('heder snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
