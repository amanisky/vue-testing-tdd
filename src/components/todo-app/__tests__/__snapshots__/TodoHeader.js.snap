// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoHeader heder snapshot 1`] = `
<header class="header">
  <h1>待办</h1> <input data-testid="new-todo" autofocus="autofocus" placeholder="需要做什么？" class="new-todo">
</header>
`;
